canvas_sdk/data/client.py

```diff
@@ -0,0 +1,68 @@
+from typing import Any, cast
+
+from gql import Client, gql
+from gql.transport.aiohttp import AIOHTTPTransport
+
+from settings import DAL_GRAPHQL_AUTH_KEY, DAL_GRAPHQL_ENDPOINT
+
+
+class _CanvasGQLClient:
+    """
+    This is a GraphQL client that can be used to query home-app in order to fetch data for use in plugins.
+
+    Usage Examples:
+
+    A query with no parameters:
+
+    TEST_QUERY_NO_PARAMS = '''
+      query PatientsAll {
+        patients {
+          edges {
+            node {
+              firstName
+              lastName
+              birthDate
+            }
+          }
+        }
+      }
+    '''
+
+    client = _CanvasGQLClient()
+    result = client.query(TEST_QUERY_NO_PARAMS)
+    print(result) # returns dictionary
+
+    A query with parameters:
+
+    TEST_QUERY_WITH_PARAMS = '''
+        query PatientGet($patientKey: String!) {
+          patient(patientKey: $patientKey) {
+            firstName
+            lastName
+            birthDate
+          }
+        }
+    '''
+
+    client = _CanvasGQLClient()
+    result = client.query(TEST_QUERY_NO_PARAMS)
+    print(result)
+    """
+
+    def __init__(self) -> None:
+        transport = AIOHTTPTransport(
+            url=cast(str, DAL_GRAPHQL_ENDPOINT),
+            headers={"Authorization": f"{cast(str, DAL_GRAPHQL_AUTH_KEY)}"},
+        )
+        self.client = Client(transport=transport, fetch_schema_from_transport=True)
+
+    def query(self, gql_query: str, variables: dict[str, Any] | None = None) -> dict[str, Any]:
+        if variables is None:
+            query_variables = {}
+        else:
+            query_variables = variables
+
+        return self.client.execute(gql(gql_query), variable_values=query_variables)
+
+
+GQL_CLIENT = _CanvasGQLClient()
```

Comment:
One more piece of feedback here -- if we want to allow clients to import the client from `canvas_sdk.data`, then we need to import the `GQL_CLIENT` into into the `__init__.py` for the data module.

Reply:
working on the JWT piece, thinking about options for incorporating the auth:

- let people import the client directly and pass a JWT to the plugin in the `secrets` object which they can use to instantiate the client
- add a method to `BaseHandler` to generate a JWT for the plugin at runtime (I worry this could be abused as the code would be accessible to plugins)
- pass a pre-configured client to the plugin by making BaseHandler accept it as an argument

I think the last one is safest and also easiest for developers to use since it will be available and pre-configured without them having to import it... what do you think?

Reply:
For the last option, are you thinking a pre-configured client would mean that each run of a plugin gets its own client object?

Or would each run of a plugin use the same underlying client object, but use the JWT for that plugin at request time?

If it's the former, then I think we would lose out on reusing the same TCP connection for all requests between the Plugin Runner and home-app.

If it's the latter, then things might get weird from a thread safety perspective, depending on how the GraphQL client marshalls the JWTs for different plugins. Having the GraphQL client hold as little state as possible would probably be a good thing.

Something that could help here are Python [contextvars](https://docs.python.org/3/library/contextvars.html). Each async task or thread has its own context-local state, and that state can be copied down to the thread that runs `compute`. So if the JWT is created and set in the async task context as a `ContextVar`, the client could pick it up in the thread right before it issues the request.

The way contextvars are set and accessed also has the benefit that you don't have to modify any existing interfaces, like `Protocol` or the GQL client. Also, we could issue new JWT per request, if that is desireable at all, like if we want to populate claims with any request-specific data or if we want to use short expiration times.

Reply:
So I think the contextvars approach would work and would be appropriate if users were using the GQL client, but given what we discussed in the SDK meeting today, I think there's a simpler option.

Rather than making the client accessible as an attribute on the protocol object, we could also just go one step further and just have a `run_gql_query` method on `BaseProtocol`. The JWT could be passed in as a secret (your first option), and `self.run_gql_query(query, variables)` could use the singleton GQL client behind the scenes to send the query with the JWT. Simple and user-friendly.

Reply:
ahhh yes, easy and makes sense 👍 will go that route

Reply:
@beaugunderson @csande I just added the client into `BaseHandler` before reading this comment. I can remove it if need be.

Reply:
I just pushed up the changes that include the `run_gql_query` method. @beaugunderson I'm not sure what the header needs to look like in the query to home-app, so I left it out for now. Do you want to take care of that with your JWT PR?

Reply:
It should look like this:

```
Authorization: Bearer <JWT>
```

Reply:
Are we going to need to pass the JWT to the client on every execution? The `gql` library doesn't have much documentation on headers. It just shows how to define them in the transport object:
https://gql.readthedocs.io/en/latest/usage/headers.html

Reply:
That's quite a wrinkle. I'm not sure you can set headers on a per-request basis with this library, which is odd.

For what it's worth, since we're hiding the client behind a method, we could actually just do HTTP requests on a requests.Session object if we have to.

Reply:
very true... will experiment tonight

Reply:
added an implementation that I hope works but haven't had a chance to test yet--`gql` will pass `extra_args` down through to the client, which in our case then updates the `post_args` with it